### PR TITLE
Machine readable interface for tsconfig

### PR DESCRIPTION
**Pull request by orta**
I'd like to be able to pull out the information about the compiler settings section of the TSConfig in some form, which can then be used for generating an API reference. This should ideally come from the TS source code, and I can the use the TS API to pull out.

This PR shows a technique which I think shouldn't be too intrusive:

- Separate out private and public parts of the `CompilerOptions` into separate interfaces
- Use JSDoc to add single line definitions 
- Code-gen enum types when we have them from sources of truth (in this case I looked at the `libs` option)

Downside: I'm abusing JSDoc by making a new block tag `@grouping` but I didn't feel like any of the [existing tags](https://jsdoc.app) would fit. 

Does this feel like a reasonable way to keep the TSConfig reference close to the source of truth?

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -4635,55 +4635,47 @@ namespace ts {
 
     export type CompilerOptionsValue = string | number | boolean | (string | number)[] | string[] | MapLike<string[]> | PluginImport[] | ProjectReference[] | null | undefined;
 
-    export interface CompilerOptions {
-        /*@internal*/ all?: boolean;
+    // The initial string negates the work somewhat internally, yes, but it does mean that machines can pull out the rest of the values from the enum  while allowing all values
+    export type AvailableLibOptions = string | "ES5" | "ES6" | "ES2015" | "ES7" | "ES2016" | "ES2017" | "ES2018" | "ESNext" | "DOM" | "DOM.Iterable" | "WebWorker" | "ScriptHost" | "ES2015.Core" | "ES2015.Collection" | "ES2015.Generator" | "ES2015.Iterable" | "ES2015.Promise" | "ES2015.Proxy" | "ES2015.Reflect" | "ES2015.Symbol" | "ES2015.Symbol.WellKnown" | "ES2016.Array.Include" | "ES2017.object" | "ES2017.Intl" | "ES2017.SharedMemory" | "ES2017.String" | "ES2017.TypedArrays" | "ES2018.Intl" | "ES2018.Promise" | "ES2018.RegExp" | "ESNext.AsyncIterable" | "ESNext.Array" | "ESNext.Intl" | "ESNext.Symbol";
```
Comment (ajafff) [inline src/compiler/types.ts]: This will reduce to `string`

Comment (orta) [reply src/compiler/types.ts]: Yeah, thanks! This is why I wrote the comment above - it's parseable out via tools for the site, even if we don't get the safety inside the compiler (where it's a bit less important anyway)

**Comment by weswigham** [comment]
@orta We already have a machine readable store of command metadata - it's how `--init` and `--help` work.  `commandLineParser.ts` has all option definitions, including descriptions. The interface within `types` is just a collection derived from those definitions plus some internal bits. It also has the list of all libs.

**Comment by orta** [comment]
Thanks!

**Pull request closed without merging**